core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -863,13 +863,15 @@ class LogManager(logDirs: Seq[File],
 
   /**
    * Delete any eligible logs. Return the number of segments deleted.
-   * Only consider logs that are not compacted.
+   * Only consider logs whose cleanup policy is not 'compact,delete', as logs with
+   * that policy are handled by the LogCleanerManager and deleting them here could
+   * multiple threads to attempt to delete the same segment (see KIP-71).
    */
   def cleanupLogs() {
     debug("Beginning log cleanup...")
     var total = 0
     val startMs = time.milliseconds
-    for(log <- allLogs; if !log.config.compact) {
+    for(log <- allLogs; if !(log.config.compact && log.config.delete)) {
```

Comment:
Hmm, do you think it's cleaner if all compacted topics (regardless of whether deletion is enabled or not) are handled in `LogCleanerManager`?

Reply:
I agree with Druvil here. We want to prevent the case that a log segment is deleted by the log retention scheduler while the log cleaner thread is doing a round of cleaning. So, it seems it's better for a compacted topic to handle log-start-offset change  in the log cleaner thread. We can call log.deleteLogStartOffsetBreachedSegments in LogCleaner.cleanOrSleep(). This is similar to how a compacted + delete topic handles log retention.

Reply:
You're right, that is cleaner.